File: smartsocket/ss68/java/tutorial/lesson4/receive2.java
/*
 * Copyright (c) 1997-2006 TIBCO Software Inc.
 * All rights reserved.
 * For more information, please contact:
 * TIBCO Software Inc., Palo Alto, California, USA
 *
 * $Id: //dev/packages/examples/ss/dev/68/src/java/tutorial/lesson4b/receive2.java#1 $
 */

//----------------------------------------------------------------------
// receive2.java -- output a NUMERIC_DATA message with a callback

import java.io.*;
import com.smartsockets.*;

public class receive2 {

  public class receiveCb
    implements TipcProcessCb, TipcDefaultCb {

    public void process(TipcMsg msg, Object arg) {
      System.out.println("Received NUMERIC_DATA message");

      // position the field ptr to the beginning of the message
      try {
        msg.setCurrent(0);
      } catch (TipcException e) {
        Tut.fatal(e);
      } // catch

      // read the data part of the message
      try {
        String var_name;
        while (true) {
          var_name = msg.nextStr();
          double var_value;
          var_value = msg.nextReal8();
          System.out.println("Var name = " + var_name +
                             ", value = " + var_value);
        } // while
      } catch (TipcException e) { }
      // catch end-of-message-data exception, do nothing.
    } // process

    // handle() is for responding to default messages
    public void handle(TipcMsg msg, Object arg) {
        System.out.println("Receive: unexpected message type name" +
          " is <" + msg.getType().getName() + ">");
    } // handle

  } // receiveCb

  public receive2() {
    TipcMsg msg = null;

    // set the ss.project
    try {
      Tut.setOption("ss.project", "smartsockets");
      TipcSrv srv=TipcSvc.getSrv();

      // create a new receive listener and register it
      receiveCb rcb = new receiveCb();
      TipcCb rcbh = srv.addProcessCb(
                       rcb, TipcSvc.lookupMt(TipcMt.NUMERIC_DATA), srv);
      // check the 'handle' returned for validity
      if (null == rcbh) {
        Tut.exitFailure("Couldn't register process listener!\n");
      }

      // register receiveCb again as a default listener
      TipcCb dcbh = srv.addDefaultCb(rcb, srv);
      // check the 'handle' returned for validity
      if (null == dcbh) {
        Tut.exitFailure("Couldn't register default listener!\n");
      }

      // connect to RTserver
      if (!srv.create()) {
        Tut.exitFailure("Couldn't connect to RTserver!\n");
      }
      // subscribe to the appropriate subject
      srv.setSubjectSubscribe("/ss/tutorial/lesson4", true);

      // read and process all incoming messages
      while (null != (msg = srv.next(TipcDefs.TIMEOUT_FOREVER))) {
        srv.process(msg);
      } // while

      // disconnect from RTserver
      srv.destroy();

      // unregister the listeners for completeness
      srv.removeProcessCb(rcbh);
      srv.removeDefaultCb(dcbh);
    } catch (TipcException e) {
      Tut.fatal(e);
    } // catch
  } // receive2 (constructor)

  public static void main(String[] argv) {
    receive2 r = new receive2();
  } // main

} // receive2 class

File: smartsocket/ss68/examples/smrtsock/cxxtutrl/lesson6/receive2.cxx
/*
 * Copyright (c) 1991-2006 TIBCO Software Inc.
 * All rights reserved.
 * For more information, please contact:
 * TIBCO Software Inc., Palo Alto, California, USA
 *
 * $Id: //dev/packages/examples/ss/dev/68/src/c/cxxtutrl/lesson6/receive2.cxx#1 $
 */

// receive2.cxx - print out contents of MSG_COUNT messages
// $RTHOME/examples/smrtsock/cxxtutrl/lesson6/receive2.cxx 

#define MSG_COUNT 2000
#include <iostream.h>
#include <rtworks/cxxipc.hxx>

// ============================================================= 
//..ProcessMsgCount - callback for processing MSG_COUNT messages 

void T_ENTRY ProcessMsgCount( T_IPC_CONN conn,
		      T_IPC_CONN_PROCESS_CB_DATA data,
		      T_CB_ARG arg )

{
  TipcMsg msg = data->msg;
  T_INT4 msg_count;

  msg.Current(0);
  msg >> msg_count;
  cout << "Message Count = " << msg_count << endl;

 } // ProcessMsgCount 

// ==================================================================== 

int main(int argc, char **argv)
{
  TipcMt mt;
  TipcMsg msg;
  TipcSrv& rtServer = TipcSrv::Instance();

  // Execute commands in file 
  TutCommandParseFile((T_STR)"receive2.cm");

  // Define a new message type, MSG_COUNT 
  mt.Create((T_STR)"msg_count", MSG_COUNT, (T_STR)"int4");

  // Connect to the RTserver 
  if (!rtServer.Create(T_IPC_SRV_CONN_FULL)) {
    cout << "Could not connect to RTserver!" << endl;
    TutExit(T_EXIT_FAILURE);
  } // if 

  // Setup callback to catch MSG_COUNT messages 
  mt.Lookup(MSG_COUNT);
  rtServer.ProcessCbCreate(mt, ProcessMsgCount, NULL);

  // Start receiving the "/tutorial/lesson6" subject 
  rtServer.SubjectSubscribe((T_STR)"/tutorial/lesson6", TRUE);

  // Read and process all incoming messages 
  while (1) {
  rtServer.MainLoop(T_TIMEOUT_FOREVER);
  }

  return T_EXIT_SUCCESS;
} // main 
